Major revamp

chrisjsewell opened the pull request.

chrisjsewell commented:
@choldgraf I done a major revamp of this, to more closely follow the bootstrap 4 system(s). Have a look at the docs and see what you think.

Note I'm using the RTD (non-bootstrap) theme here, because I wanted to check that it works for the AiiDA docs (the initial use case). But ideally we could make the docs for both this theme and the pydata one, like nbsphinx does with their docs and multiple theme builds: https://nbsphinx.readthedocs.io

chrisjsewell commented on sphinx_panels/__init__.py:
```diff
@@ -70,8 +166,32 @@ def add_static_path(app):
+
     app.add_css_file("panels.css")
+
+    # TODO only load these is using a non-boostrap theme
+    app.add_css_file("bs-grids.css")
+    app.add_css_file("bs-cards.css")
```
Here I'd like to find a way to only load these CSS if `bootstrap.min.css` is not already being loaded (they just contain copied CSS from that, specific to each element)

chrisjsewell replied:
I can't see any easy way to automate bootstrap theme discovery so, in e67f3ac, I have added a sphinx option to disable loading the bootstrap CSS

choldgraf commented on docs/index.rst:
```diff
@@ -129,54 +130,132 @@ Card Layout
+------------
+
+To style the look of cards,
+you may use the directive options to add default CSS classes for eac element,
```
```suggestion
you may use the directive options to add default CSS classes for each element,
```

choldgraf commented on docs/index.rst:
```diff
@@ -6,8 +6,8 @@ A sphinx extension for creating panels in a grid layout.
 utilising both the bootstrap 4
-`grid system <https://www.w3schools.com/bootstrap/bootstrap_grid_system.asp>`_,
-and `cards layout <https://www.w3schools.com/bootstrap4/bootstrap_cards.asp>`_.
+`grid system <https://getbootstrap.com/docs/4.0/layout/grid/>`_,
```
Is there any reason to think we'd run into issues if other bootstrap CSS/JS gets loaded on top of the CSS rules in here? (other than some unnecessary data transfer)

chrisjsewell replied:
This is why you can set `add_boostrap_css = False` in `conf.py`, then no CSS will be loaded by this extension. Then there should be no issue?

choldgraf replied:
ah right - I agree. Can we add a short prefix to that configuration value? E.g. `panels_add_bootstrap_css`, as Sphinx suggests?

chrisjsewell replied:
fixed in 0a312cd

choldgraf replied:
another thought here is that if we had a "meta" extension that loaded bootstrap, that extension could be configured selectively in a single variable, rather than in a bunch of one-off configurations. e.g.:

```
sphinx_bootstrap_load_css = ["panel", "grid"]
```

etc. Or just `"all"` for all CSS classes

choldgraf commented on docs/index.rst:
```diff
@@ -93,12 +93,13 @@ then panel specific classes can be added at the start of each panel.
 
     .. panels::
-        :container: container-lg pad-bottom-20
+        :container: container-lg pb-3
```
This doesn't *always* need `:container:` though, right? I feel like it should have reasonable defaults so users don't have to start writing their own bootstrap classes off the bat

chrisjsewell replied:
No, the defaults are described in the lines above this, and are defined in the code here: https://github.com/executablebooks/sphinx-panels/blob/195e853a214df6234c107d5015b3213070bec42c/sphinx_panels/__init__.py#L34-L36

Note that all classes I use in the documentation are built-in boostrap classes, which are included in the CSS loaded here

choldgraf commented on docs/index.rst:
```diff
@@ -129,54 +130,132 @@ Card Layout
+but it is advised you use the built-in bootstrap classes:
+
+- `Card colouring <https://getbootstrap.com/docs/4.0/utilities/colors/>`_  contextual classes: `bg-primary`, `bg-success`, `bg-info`, `bg-warning`, `bg-danger`, `bg-secondary`, `bg-dark` and `bg-light`.
+- `Padding and margins <https://getbootstrap.com/docs/4.0/utilities/spacing/>`_: `border-0`, `p-2`, `m-2`, ...
```
This makes me wonder if there'd be value in having something like "sphinx-bootstrap" that is just an easy way to define a few useful bootstrap elements with sphinx directives. You could imagine instructions like these being applied to a number of UI elements. (as opposed to having one sphinx extension per element type, such as "sphinx-panels", "sphinx-togglebutton", etc)

chrisjsewell replied:
yep, if this existed it would be easy to add the dependency to this package. If you look at the CSS I have added, you'll see that I have divided into approximately the classes required for each aspect.

choldgraf replied:
so there'd be something like a bootstrap meta-package that would make it possible for extension to depend on it and then those extensions don't have to worry about loading any bootstrap JS/CSS themselves?

That might be nice, then we could look into re-working `sphinx-togglebutton` so that it uses bootstrap, but without requiring users to use a theme that has bootstrap

choldgraf replied:
Another thing that could benefit from bootstrap would be "question and answer" blocks: https://getbootstrap.com/docs/4.1/components/forms/

choldgraf commented on docs/index.rst:
```diff
@@ -129,54 +130,132 @@ Card Layout
+    :footer: text-right
+
+    ---
+    column += p-1
```
It's unclear what's happening here - are you running your own python code inside of the panel content?

chrisjsewell replied:
just using regexes to identify <column><+=><p-1>, then deciding whether to add to the existing classes, or reset them based on if `=` or `+=`. See: https://github.com/executablebooks/sphinx-panels/blob/195e853a214df6234c107d5015b3213070bec42c/sphinx_panels/__init__.py#L134-L139

choldgraf commented on setup.py:
```diff
@@ -24,18 +24,19 @@
     packages=find_packages(),
     package_data={
-        "sphinx_panels": ["css/panels.css", "css/bs-cards.css", "css/bs-grids.css"]
+        "sphinx_panels": ["css/bs-cards.css", "css/bs-grids.css", "css/bs-borders.css"]
```
I take it that these are straight from the bootstrap bundle?

chrisjsewell replied:
yes all the css they contain comes from `boostrap.min.css`, but split into loosely the functionality they apply to

choldgraf replied:
gotcha - so if bootstrap updates itself and we want those updates, we need to find the right part of `.min.css`, then copy paste those sections and put them in these files?

choldgraf commented on sphinx_panels/__init__.py:
```diff
@@ -60,114 +184,70 @@ def run(self):
-                title = nodes.container(
-                    in_panel=True, classes=["card-title"] + classes["title"]
+            if "header" in data:
+                header = nodes.container(
```
one "gotcha" that we should make sure isn't an issue here is the word `container`. the pydata them realized that weird things were happening because `container` is a very common and specific bootstrap class, while Sphinx appends `container` to lots of things as well

chrisjsewell replied:
I specifically account for this in: https://github.com/executablebooks/sphinx-panels/blob/195e853a214df6234c107d5015b3213070bec42c/sphinx_panels/__init__.py#L302-L307

chrisjsewell commented:
Thanks for the comments @choldgraf, I'll look now. Note in 16794f4 I have just changed the default delimiters (see docs), so that they are less likely to clash with rST header underlines, but also added a config value to change them. Open to suggestions on what the defaults should be though.

choldgraf commented:
re: the delimiters, I'm guessing you've already considered and decided against just having another directive for these? E.g. following what sphinx-tabs does:

```
.. tabs::
   :tabs: config

   .. tab::
      :singletab: config

      tab1 content
   
   .. tab::
      :singletab: config

      tab2 content

   .. tab::
      :singletab: config

      tab3 content
```

?

chrisjsewell commented:
> re: the delimiters, I'm guessing you've already considered and decided against just having another directive for these?

Well yeh, I feel that (a) it adds the need for extra indentation (in rST at least), which I don't think is ideal and (b) it implies, assuming a `panel` directive, that it can be outside the `panels` directive, and/or that there can be content in-between the `panel` directives, neither of which really make sense.

choldgraf commented:
Yeah that makes sense - I think that the pros and cons of parsing panel structure/config with a single directive basically comes down to complexity of the panels. The more complex you want your panel structure to be (or, the more you want it to deviate from the default), the more clunky it'll feel to configure with special syntax that only applies to the `sphinx-panels` directive. The less-complex you want it to be, the more it'll be straightforward to just use `---` delimiters or whatever.

chrisjsewell commented:
> The more complex you want your panel structure to be

But then you can just use the nested `container` directive approach that pandas-dev/pandas#32065 uses. In some respect, this is meant to be a short-hand for that, which is simpler to type and doesn't leave you with large amounts of nested directives in your source code.

choldgraf commented:
@chrisjsewell that's a good point...at that point you're a "power user" so it's probably fine to expect folks to build their own containers etc

chrisjsewell commented:
> at that point you're a "power user"

Personally, as a "power user", I would still use this and feel it's a lot less clunky than having to use multiple directives. But yep, different strokes for different folks.

Ok cool well I think this is good to merge and release as v0.2.0. Unless you want to discuss anything else?

choldgraf commented:
let's do it 🚀 - let's keep this "meta bootstrap" repository idea on the backburner and see if it could be useful in the future. I really like the idea of finding a way to expose bootstrap functionality to sphinx users without requiring them to use a particular theme

choldgraf commented:
You know right after I pressed merge I had one other thought - rather than using a delimiter that's also got meaning in markdown/rST, what if we just used `{panel}` as a delimiter?

e.g.:


```
.. panels::

    Content of the top-left panel

    {panel}

    Content of the top-right panel

    {panel}

    Content of the bottom-left panel

    {panel}

    Content of the bottom-right panel
```

chrisjsewell commented:
eurgh, dunno I prefer being able to use variable length delimiters. I think it's easier to see the delimiters quickly, if you had a lot of content in the panels, with:

```rst
.. panels::

    Content of the top-left panel

    ...................................................

    Content of the top-right panel

    ...................................................

    Content of the bottom-left panel
```

Happy to put it out for poll or something though lol

choldgraf commented:
nah I think it's fine - I hadn't thought about variable-length. was just an idea to explore if you immediately jumped on it, but if your first thought is hesitation I think we should just go w/ the current approach and see if it raises problems over time

choldgraf commented:
One other thought, then - the way that sphinx-gallery does this is to use two options for people. One is delimiters with `#`. e.g.:

```
Block 1

###################

Block 2
```

Which has the added benefit that it doesn't overlap with rST and also won't clash with ATX headers in markdown

Or another option which is to use Hydrogen and Pycharm-style delimiters like so:

```
Block 1

# %%

Blocks 2
```

I wonder if it would be good to piggy-back on one of those since there is already some prior art in the community for using those as delimiters?

chrisjsewell commented:
Using `#` as a delimiter would just be a matter of changing the default in: https://github.com/executablebooks/sphinx-panels/blob/b8d6b36f946a6d86b264323189866c5e48eeff0b/sphinx_panels/__init__.py#L316

What is the link where you can do this in sphinx-gallery I can't actually find it

choldgraf commented:
Here's where they describe how to structure sphinx-gallery python files:

https://sphinx-gallery.github.io/stable/syntax.html#embed-rst-in-your-example-python-files

They have a different use-case, which is to break up blocks of **python** into content blocks and code blocks, but perhaps the same idea could be used here?

chrisjsewell commented:
ok right but (to state the obvious) thats for using rST in python files, which is not really the same use case. TBH I think there is enough flexibility in making the delimiters configurable. The only thing I would think of adding, is to allow the header and footer to be `None` to "switch them off", then you would only have one potential piece of syntax to clash with any existing syntax
